File: app/modules/tulmi-stream/ios/TulmiStreamModule.swift
import ExpoModulesCore
import AVFoundation

/// Live (streaming) dictation for the main app.
///
/// JS calls `start({ url, token, targetApp, language })`; the native side opens a
/// WebSocket to the backend, captures the mic as 16 kHz mono PCM, streams it, and
/// emits `onReady` / `onPartial` / `onFinal` / `onError` / `onClosed` events back
/// to JS. See STREAMING.md for the wire protocol.
public class TulmiStreamModule: Module {
  private var streamer: Streamer?

  public func definition() -> ModuleDefinition {
    Name("TulmiStream")

    Events("onReady", "onPartial", "onFinal", "onError", "onClosed")

    Function("start") { (options: [String: Any]) in
      let url = options["url"] as? String ?? ""
      let token = options["token"] as? String ?? "dev"
      let targetApp = options["targetApp"] as? String ?? "Generic"
      let language = options["language"] as? String ?? "auto"
      // The caller says whether this session also has to PLAY. Dictation does
      // not and keeps the narrower .record category; the spoken conversation
      // does, and a .record session cannot answer.
      let duplex = options["duplex"] as? Bool ?? false
      self.streamer?.cancel()
      let s = Streamer(duplex: duplex) { [weak self] name, payload in
        self?.sendEvent(name, payload)
      }
      self.streamer = s
      s.start(urlString: url, token: token, targetApp: targetApp, language: language)
    }

    Function("stop") {
      self.streamer?.finish()
    }

    Function("cancel") {
      self.streamer?.cancel()
      // A duplex streamer holds the audio session between turns, so leaving the
      // screen has to hand it back explicitly — otherwise the app keeps the
      // route to itself and whatever the user was playing stays silenced.
      self.streamer?.releaseSession()
      self.streamer = nil
    }

    OnDestroy {
      self.streamer?.cancel()
      self.streamer = nil
    }
  }
}

/// The actual capture + WebSocket plumbing. Mirrors the keyboard's TulmiStream,
/// but reports through an event closure instead of an enum callback.
private final class Streamer: NSObject {
  private let emit: (String, [String: Any]) -> Void
  /// True when this session also plays — see activateSession().
  private let duplex: Bool
  private let session = URLSession(configuration: .default)
  private var task: URLSessionWebSocketTask?

  private let engine = AVAudioEngine()
  private var converter: AVAudioConverter?
  private let targetFormat = AVAudioFormat(
    commonFormat: .pcmFormatInt16, sampleRate: 16000, channels: 1, interleaved: true
  )!
  private var tapInstalled = false

  /// Guards the object refs the realtime audio tap reads (task/converter) while
  /// the main thread frees them — reading an object reference mid-write is
  /// undefined in Swift and can over-release → use-after-free. Same pattern as
  /// FlowSessionManager; the tap holds it only for a two-pointer snapshot.
  private let avLock = NSLock()
  private func captureAV() -> (AVAudioConverter?, URLSessionWebSocketTask?) {
    avLock.lock(); defer { avLock.unlock() }; return (converter, task)
  }
  private func setTask(_ t: URLSessionWebSocketTask?) { avLock.lock(); task = t; avLock.unlock() }
  private func setConverter(_ c: AVAudioConverter?) { avLock.lock(); converter = c; avLock.unlock() }

  init(duplex: Bool, emit: @escaping (String, [String: Any]) -> Void) {
    self.duplex = duplex
    self.emit = emit
    super.init()
  }

  func start(urlString: String, token: String, targetApp: String, language: String) {
    guard let url = URL(string: urlString) else {
      emit("onError", ["message": "Bad server URL"])
      return
    }
    var req = URLRequest(url: url)
    req.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
    let task = session.webSocketTask(with: req)
    setTask(task)
    task.resume()
    receiveLoop()

    let start: [String: Any] = [
      "type": "start",
      "token": token,
      "targetApp": targetApp,
      "language": language,
      "sampleRate": 16000,
      "encoding": "pcm_s16le",
      "channels": 1,
    ]
    if let data = try? JSONSerialization.data(withJSONObject: start),
       let str = String(data: data, encoding: .utf8) {
      task.send(.string(str)) { _ in }
    }

    startCapture()
  }

  func finish() {
    // Keep the socket open after "stop" so the engine's flushed tail + "done"
    // still arrive (cancelling here truncated the ending). Watchdog force-closes
    // if "done" never comes.
    stopCapture()
    guard let task = task else { emit("onClosed", [:]); return }
    task.send(.string("{\"type\":\"stop\"}")) { _ in }
    DispatchQueue.main.asyncAfter(deadline: .now() + 2.5) { [weak self] in
      guard let self = self, self.task != nil else { return }
      self.task?.cancel(with: .normalClosure, reason: nil)
      self.setTask(nil)
      self.emit("onClosed", [:])
    }
  }

  func cancel() {
    stopCapture()
    task?.cancel(with: .goingAway, reason: nil)
    setTask(nil)
  }

  private func startCapture() {
    if !activateSession() { return }
    let input = engine.inputNode
    let inputFormat = input.outputFormat(forBus: 0)
    setConverter(AVAudioConverter(from: inputFormat, to: targetFormat))
    input.installTap(onBus: 0, bufferSize: 2048, format: inputFormat) { [weak self] buffer, _ in
      self?.sendBuffer(buffer, inputFormat: inputFormat)
    }
    tapInstalled = true
    engine.prepare()
    do {
      try engine.start()
    } catch {
      // Don't leak the tap + active audio session when the engine won't start.
      stopCapture()
      emit("onError", ["message": "Mic start: \(error.localizedDescription)"])
    }
  }

  /// Take the audio session, and say what for.
  ///
  /// `.record` IS THE WRONG CATEGORY WHEN THE APP ALSO TALKS. A recording
  /// session cannot play, so on the spoken-conversation screen — which listens,
  /// then answers through the synthesiser, then listens again — the two fight
  /// over the same session every turn. The synthesiser holds it to speak, the
  /// next `setActive(true)` arrives while it is still letting go, and iOS
  /// answers "session activation failed". Which is what the screen showed.
  ///
  /// `.playAndRecord` is one session that serves both, so there is nothing to
  /// hand back and forth. `.defaultToSpeaker` because that category otherwise
  /// routes playback to the earpiece, and an assistant that can only be heard
  /// by holding the phone to your head is not one anybody would use.
  ///
  /// Dictation keeps `.record`: it is the narrower permission, it never plays
  /// anything, and changing the category under it would change the input path
  /// for the feature that matters most.
  private func activateSession() -> Bool {
    let audio = AVAudioSession.sharedInstance()
    let category: AVAudioSession.Category = duplex ? .playAndRecord : .record
    let options: AVAudioSession.CategoryOptions =
      duplex ? [.defaultToSpeaker, .allowBluetooth, .allowBluetoothA2DP] : []
    do {
      try audio.setCategory(category, mode: .default, options: options)
      try audio.setActive(true)
      return true
    } catch {
      // ONE RETRY, because the usual cause is a race rather than a refusal:
      // whatever held the session a moment ago is still releasing it, and by
      // the time a person could read an error it would have worked. Failing
      // twice is a real failure and is reported as one.
      Thread.sleep(forTimeInterval: 0.12)
      do {
        try audio.setCategory(category, mode: .default, options: options)
        try audio.setActive(true)
        return true
      } catch {
        emit("onError", ["message": "Audio session: \(error.localizedDescription)"])
        return false
      }
    }
  }

  private func stopCapture() {
    if tapInstalled {
      engine.inputNode.removeTap(onBus: 0)
      tapInstalled = false
    }
    if engine.isRunning { engine.stop() }
    // HOLD THE SESSION IN DUPLEX. The conversation stops the mic between every
    // turn so it does not transcribe its own reply, and dropping the session
    // there means the synthesiser has to take it and give it back on each one —
    // which is the handover that fails. Keeping it means the turn is just a tap
    // being removed.
    //
    // Dictation still releases it, and does so politely: without
    // notifyOthersOnDeactivation whatever was playing before is left paused.
    if !duplex {
      try? AVAudioSession.sharedInstance().setActive(false, options: .notifyOthersOnDeactivation)
    }
  }

  /// Release the session for good. Only the conversation needs this, because
  /// only the conversation holds on between turns.
  func releaseSession() {
    try? AVAudioSession.sharedInstance().setActive(false, options: .notifyOthersOnDeactivation)
  }

  private func sendBuffer(_ buffer: AVAudioPCMBuffer, inputFormat: AVAudioFormat) {
    // Snapshot the object refs under the lock so main can't free them mid-use.
    let (snapConv, snapTask) = captureAV()
    guard let converter = snapConv, let task = snapTask else { return }
    let ratio = targetFormat.sampleRate / inputFormat.sampleRate
    let capacity = AVAudioFrameCount(Double(buffer.frameLength) * ratio + 1024)
    guard let out = AVAudioPCMBuffer(pcmFormat: targetFormat, frameCapacity: capacity) else { return }
    var fed = false
    var err: NSError?
    let status = converter.convert(to: out, error: &err) { _, outStatus in
      if fed {
        outStatus.pointee = .noDataNow
        return nil
      }
      fed = true
      outStatus.pointee = .haveData
      return buffer
    }
    guard status != .error, out.frameLength > 0, let ch = out.int16ChannelData else { return }
    let data = Data(bytes: ch[0], count: Int(out.frameLength) * MemoryLayout<Int16>.size)
    task.send(.data(data)) { _ in }
  }

  private func receiveLoop() {
    task?.receive { [weak self] result in
      guard let self = self else { return }
      switch result {
      case .failure:
        self.emit("onClosed", [:])
      case .success(let message):
        switch message {
        case .string(let text): self.handleMessage(text)
        case .data(let data): self.handleMessage(String(data: data, encoding: .utf8) ?? "")
        @unknown default: break
        }
        self.receiveLoop()
      }
    }
  }

  private func handleMessage(_ text: String) {
    guard
      let data = text.data(using: .utf8),
      let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
      let type = json["type"] as? String
    else { return }
    switch type {
    case "ready": emit("onReady", [:])
    case "partial": emit("onPartial", ["text": json["text"] as? String ?? ""])
    case "final": emit("onFinal", ["text": json["text"] as? String ?? ""])
    // "done" is the terminal marker, not a transcript — no text to insert.
    case "done": emit("onClosed", [:])
    case "error": emit("onError", ["message": json["message"] as? String ?? "stream error"])
    default: break
    }
  }
}

File: app/modules/tulmi-stream/android/src/main/java/expo/modules/tulmistream/TulmiStreamModule.kt
package expo.modules.tulmistream

import android.media.AudioFormat
import android.media.AudioRecord
import android.media.MediaRecorder
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.Response
import okhttp3.WebSocket
import okhttp3.WebSocketListener
import okio.ByteString
import okio.ByteString.Companion.toByteString
import org.json.JSONObject
import java.util.concurrent.TimeUnit
import kotlin.concurrent.thread

/**
 * Live (streaming) dictation for the main app.
 *
 * JS calls `start({ url, token, targetApp, language })`; the native side opens a
 * WebSocket, captures the mic as 16 kHz mono PCM, streams it, and emits
 * `onReady` / `onPartial` / `onFinal` / `onError` / `onClosed` back to JS.
 * See STREAMING.md for the wire protocol.
 */
class TulmiStreamModule : Module() {
  private var streamer: Streamer? = null

  override fun definition() = ModuleDefinition {
    Name("TulmiStream")

    Events("onReady", "onPartial", "onFinal", "onError", "onClosed")

    Function("start") { options: Map<String, Any?> ->
      val url = options["url"] as? String ?: ""
      val token = options["token"] as? String ?: "dev"
      val targetApp = options["targetApp"] as? String ?: "Generic"
      val language = options["language"] as? String ?: "auto"
      streamer?.cancel()
      streamer = Streamer { name, payload -> sendEvent(name, payload) }.also {
        it.start(url, token, targetApp, language)
      }
    }

    Function("stop") { streamer?.finish() }

    Function("cancel") {
      streamer?.cancel()
      streamer = null
    }

    OnDestroy {
      streamer?.cancel()
      streamer = null
    }
  }
}

/** Capture + WebSocket plumbing; mirrors the keyboard's Stream.kt. */
private class Streamer(
  private val emit: (String, Map<String, Any?>) -> Unit,
) {
  private val client = OkHttpClient.Builder()
    .readTimeout(0, TimeUnit.MILLISECONDS) // keep the socket open
    .build()

  private var ws: WebSocket? = null
  private var record: AudioRecord? = null
  private var captureThread: Thread? = null
  @Volatile private var capturing = false

  // Backpressure cap — mirrors the keyboard's Stream.kt sendCapBytes. On a poor
  // uplink an unbounded OkHttp send queue balloons until the process OOMs; we
  // drop the newest frames once queued bytes exceed the cap and log the first
  // drop only so a bad network doesn't flood logcat.
  private val sendCapBytes = 2 * 1024 * 1024
  @Volatile private var dropLogged = false

  fun start(url: String, token: String, targetApp: String, language: String) {
    val req = Request.Builder()
      .url(url)
      .addHeader("Authorization", "Bearer $token")
      .build()
    ws = client.newWebSocket(req, object : WebSocketListener() {
      override fun onOpen(webSocket: WebSocket, response: Response) {
        val start = JSONObject()
          .put("type", "start")
          .put("token", token)
          .put("targetApp", targetApp)
          .put("language", language)
          .put("sampleRate", 16000)
          .put("encoding", "pcm_s16le")
          .put("channels", 1)
        webSocket.send(start.toString())
        startCapture(webSocket)
      }

      override fun onMessage(webSocket: WebSocket, text: String) = handle(text)
      override fun onMessage(webSocket: WebSocket, bytes: ByteString) = handle(bytes.utf8())

      override fun onFailure(webSocket: WebSocket, t: Throwable, response: Response?) {
        stopCapture()
        emit("onError", mapOf("message" to (t.message ?: "stream failed")))
      }

      override fun onClosed(webSocket: WebSocket, code: Int, reason: String) {
        stopCapture()
        emit("onClosed", emptyMap())
      }
    })
  }

  private fun handle(text: String) {
    try {
      val o = JSONObject(text)
      when (o.optString("type")) {
        "ready" -> emit("onReady", emptyMap())
        "partial" -> emit("onPartial", mapOf("text" to o.optString("text")))
        "final" -> emit("onFinal", mapOf("text" to o.optString("text")))
        // "done" carries no text — it's the terminal marker, not a transcript.
        "done" -> emit("onClosed", emptyMap())
        // A mid-stream backend error is terminal: release the mic immediately so
        // it doesn't stay hot if the JS onError handler forgets to call cancel().
        "error" -> {
          stopCapture()
          emit("onError", mapOf("message" to o.optString("message", "stream error")))
        }
      }
    } catch (_: Exception) { /* ignore malformed frames */ }
  }

  @Suppress("MissingPermission") // the app requests RECORD_AUDIO before starting
  private fun startCapture(webSocket: WebSocket) {
    val minBuf = AudioRecord.getMinBufferSize(
      16000, AudioFormat.CHANNEL_IN_MONO, AudioFormat.ENCODING_PCM_16BIT
    )
    val bufSize = maxOf(minBuf, 4096)
    val rec = try {
      AudioRecord(
        MediaRecorder.AudioSource.MIC,
        16000,
        AudioFormat.CHANNEL_IN_MONO,
        AudioFormat.ENCODING_PCM_16BIT,
        bufSize * 2,
      )
    } catch (e: SecurityException) {
      emit("onError", mapOf("message" to "Microphone permission denied"))
      return
    }
    if (rec.state != AudioRecord.STATE_INITIALIZED) {
      rec.release()
      emit("onError", mapOf("message" to "Mic unavailable"))
      return
    }
    record = rec
    capturing = true
    rec.startRecording()
    captureThread = thread(name = "tulmi-mic") {
      val buf = ByteArray(bufSize)
      while (capturing) {
        val n = rec.read(buf, 0, buf.size)
        if (n <= 0) continue
        // Drop newest frames when the socket is backlogged rather than queuing
        // forever (the process would OOM on a slow uplink). Matches the keyboard.
        if (webSocket.queueSize() + n > sendCapBytes) {
          if (!dropLogged) {
            dropLogged = true
            android.util.Log.w(
              "TulmiStream",
              "backpressure: dropping audio frames (queueSize=${webSocket.queueSize()}, cap=$sendCapBytes)",
            )
          }
          continue
        }
        webSocket.send(buf.toByteString(0, n))
      }
    }
  }

  @Synchronized
  private fun stopCapture() {
    // Idempotent — finish()/cancel()/onClosed/onError can all land here from
    // different threads; @Synchronized + this guard stop a double stop/release.
    if (record == null && captureThread == null) return
    capturing = false
    // Snapshot + detach so a concurrent caller can't touch the same refs.
    val t = captureThread; captureThread = null
    val rec = record; record = null
    // Release on a teardown thread AFTER an UNBOUNDED join. The mic thread blocks
    // in rec.read(); it exits only once that read returns (capturing is now
    // false). The OLD bounded join(700) released the AudioRecord even if the
    // join timed out — freeing native mic memory out from under an in-flight
    // read = use-after-free crash. An unbounded join guarantees the read has
    // returned before release; off the caller thread so stop() doesn't block.
    thread(name = "tulmi-mic-teardown") {
      try { t?.join() } catch (_: InterruptedException) {}
      try { rec?.stop() } catch (_: Exception) {}
      try { rec?.release() } catch (_: Exception) {}
    }
  }

  fun finish() {
    // Keep the socket open after "stop" so the engine's flushed tail + "done"
    // still arrive (closing here truncated the ending). Watchdog force-closes
    // if "done" never comes.
    stopCapture()
    val socket = ws ?: run { emit("onClosed", emptyMap()); return }
    socket.send("{\"type\":\"stop\"}")
    android.os.Handler(android.os.Looper.getMainLooper()).postDelayed({
      try { ws?.close(1000, null) } catch (_: Exception) {}
    }, 2500)
  }

  fun cancel() {
    stopCapture()
    ws?.cancel()
    ws = null
  }
}
